chirp/drivers/uvk5.py

```diff
@@ -0,0 +1,1119 @@
+TONE_RDCS=3
+
+#DTCS_CODES = sorted(chirp_common.DTCS_CODES)
+CTCSS_TONES = [
+        67.0, 69.3, 71.9, 74.4, 77.0, 79.7, 82.5, 85.4,
+        88.5, 91.5, 94.8, 97.4, 100.0, 103.5, 107.2, 110.9,
+        114.8, 118.8, 123.0, 127.3, 131.8, 136.5, 141.3, 146.2,
+        151.4, 156.7, 159.8, 162.2, 165.5, 167.9, 171.3, 173.8,
+        177.3, 179.9, 183.5, 186.2, 189.9, 192.8, 196.6, 199.5,
+        203.5, 206.5, 210.7, 218.1, 225.7, 229.1, 233.6, 241.8,
+        250.3, 254.1
+        ]
+
+#       from ft4.py
+DTCS_CODES = [ 
+              23,  25,  26,  31,  32,  36,  43,  47,  51,  53,  54,
+              65,  71,  72,  73,  74,  114, 115, 116, 122, 125, 131,
+              132, 134, 143, 145, 152, 155, 156, 162, 165, 172, 174,
+              205, 212, 223, 225, 226, 243, 244, 245, 246, 251, 252,
+              255, 261, 263, 265, 266, 271, 274, 306, 311, 315, 325,
+              331, 332, 343, 346, 351, 356, 364, 365, 371, 411, 412,
+              413, 423, 431, 432, 445, 446, 452, 454, 455, 462, 464,
+              465, 466, 503, 506, 516, 523, 526, 532, 546, 565, 606,
+              612, 624, 627, 631, 632, 654, 662, 664, 703, 712, 723,
+              731, 732, 734, 743, 754
+              ]
+
+FLOCK_LIST=[ "Off", "FCC", "CE", "GB", "430", "438" ]
+SCANRESUME_LIST = [ "TO: Resume after 5 seconds", "CO: Resume after signal dissapears", "SE: Stop scanning after receiving a signal" ]
+WELCOME_LIST = [ "Full Screen", "Welcome Info", "Voltage" ]
+KEYPADTONE_LIST = [ "Off", "Chinese", "English" ]
+LANGUAGE_LIST = [ "Chinese", "English" ]
+ALARMMODE_LIST = [ "SITE", "TONE" ]
+REMENDOFTALK_LIST = [ "Off" , "ROGER", "MDC" ]
+RTE_LIST = [ "Off" , "100ms", "200ms", "300ms", "400ms", "500ms", "600ms", "700ms", "800ms", "900ms" ]
+
+MEM_SIZE=0x2000 #size of all memory
+PROG_SIZE=0x1d00 #size of the memowy that we will program
+MEM_BLOCK=0x80 #largest block of memory that we can reliably write
+
+# the communication is obfuscated using this fine mechanism
+def xorarr(data: bytes):
+    tbl=[ 22, 108, 20, 230, 46, 145, 13, 64, 33, 53, 213, 64, 19, 3, 233, 128]
+    #x=[]
+    x=b""
+    r=0
+    for byte in data:
+        x+=bytes([byte^tbl[r]])
+        r=(r+1)%len(tbl)
+    return x
+
+# if this crc was used for communication to AND from the radio, then it
+# would be a measure to increase reliability.
+# but it's only used towards the radio, so it's for further obfuscation
+def calculate_crc16_xmodem(data: bytes):
+    poly=0x1021
+    crc = 0x0
+    for byte in data:
+        crc = crc ^ (byte << 8)
+        for i in range(8):
+            crc = crc << 1
+            if (crc & 0x10000):
+                crc = (crc ^ poly ) & 0xFFFF
+    return crc & 0xFFFF
+
+
+def _send_command(radio,data: bytes):
+    """Send a command to UV-K5 radio"""
+    serial=radio.pipe
+    serial.timeout=0.5
+
+    crc=calculate_crc16_xmodem(data)
+    data2=data+bytes([crc&0xff,(crc>>8)&0xff])
+
+    command=b"\xAB\xCD"+bytes([len(data)])+b"\x00"+xorarr(data2)+b"\xDC\xBA"
```

Comment:
Please use `struct.pack()` instead of just binary string concatenation for as much as possible. We have a lot of legacy binary-in-string stuff from the python 2 days, but I'd really prefer not to add much more.

Reply:
Fixed